### PR TITLE
[Task-15][Backend] Keyword report with google search information & [Task-16][Frontend] Keyword report with google search information

### LINKED ISSUE #15: [Task-15][Backend] Keyword report with google search information
## Why

Users should see the keywords they uploaded with reports extracted from the google page.
 
## Acceptance Criteria
- [ ] Create report with information
- [ ] Show HTML page

### LINKED ISSUE #16: [Task-16][Frontend] Keyword report with google search information
## Why

A better view of the report with pagination. 

## Acceptance Criteria

- [ ] Nice bootstrap table
- [ ] Support pagination

### PR DESCRIPTION
**Pull request by rafayet-monon**
Task Card - https://github.com/rafayet-monon/elixir_search_extractor/projects/1#card-61409759
                  - https://github.com/rafayet-monon/elixir_search_extractor/projects/1#card-61411826

## What happened 👀

Show keyword report and HTML view

## Insight 📝

- [x] Create report with information
- [x] Show HTML page
- [x] Nice bootstrap table
- [x] Support pagination

## Proof Of Work 📹

- Report
image

- HTML view
image

### CODE REVIEW

File: lib/elixir_search_extractor/search_keyword/queries/keyword_query.ex [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+  import Ecto.Query
+  alias ElixirSearchExtractor.FileUpload.Schemas.KeywordFile
+  alias ElixirSearchExtractor.SearchKeyword.Schemas.Keyword
+
+  def user_keywords(user) do
+    from keyword in Keyword,
+      join: file in KeywordFile,
+      on: keyword.keyword_file_id == file.id,
+      where: file.user_id == ^user.id,
+      order_by: fragment("? DESC", keyword.inserted_at)
```
Comment (hanam1ni) [inline lib/elixir_search_extractor/search_keyword/queries/keyword_query.ex]: I think `fragment` is not needed here to make the code simpler
```suggestion
      order_by: [desc: :inserted_at]
```

File: lib/elixir_search_extractor/search_keyword/search_keywords.ex [hunk 2]
```diff
@@ -2,9 +2,20 @@ defmodule ElixirSearchExtractor.SearchKeyword.SearchKeywords do
...
   alias ElixirSearchExtractorWorker.KeywordSearchWorker
 
+  def paginated_user_keywords(user, params \\ %{}) do
+    user
+    |> KeywordQuery.user_keywords()
+    |> Repo.paginate(params)
+  end
+
+  def get_keyword!(keyword_id) do
+    Repo.get_by!(Keyword, %{id: keyword_id})
```
Comment (hanam1ni) [inline lib/elixir_search_extractor/search_keyword/search_keywords.ex]:
```suggestion
    Repo.get!(Keyword, keyword_id)
```

File: test/elixir_search_extractor/search_keyword/search_keywords_test.exs [hunk 3]
```diff
@@ -1,10 +1,33 @@
 defmodule ElixirSearchExtractor.SearchKeyword.SearchKeywordsTest do
   use ElixirSearchExtractor.DataCase, async: true
 
-  import ElixirSearchExtractor.KeywordFixtures
+  import ElixirSearchExtractor.{AccountsFixtures, KeywordFixtures}
   alias ElixirSearchExtractor.SearchKeyword.Schemas.Keyword
   alias ElixirSearchExtractor.SearchKeyword.SearchKeywords
 
+  describe "paginated_user_keywords/2" do
+    test "returns user's keywords" do
```
Comment (hanam1ni) [inline test/elixir_search_extractor/search_keyword/search_keywords_test.exs:9]: let's add a keyword that belongs to another user to make sure the only keyword that belongs to the given user is returned.

File: test/elixir_search_extractor/search_keyword/search_keywords_test.exs [hunk 4]
```diff
@@ -1,10 +1,33 @@
...
+
+      {keyword, _} = SearchKeywords.paginated_user_keywords(user, %{page: 1})
+
+      assert length(keyword) == 1
+
+      assert List.first(keyword).id == user_keyword.id
+    end
+  end
+
+  describe "get_keyword!/1" do
```
Comment (hanam1ni) [inline test/elixir_search_extractor/search_keyword/search_keywords_test.exs:23]: let's add a case when keyword not found

File: test/elixir_search_extractor_web/controllers/keyword_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+defmodule ElixirSearchExtractorWeb.KeywordControllerTest do
+  use ElixirSearchExtractorWeb.ConnCase
```
Comment (hanam1ni) [inline test/elixir_search_extractor_web/controllers/keyword_controller_test.exs]:
```suggestion
  use ElixirSearchExtractorWeb.ConnCase, async: true
```

**Comment by rafayet-monon** [comment]
@hanam1ni Fixed your reviews in [c330255](https://github.com/rafayet-monon/elixir_search_extractor/pull/41/commits/c330255356218d670f9f08a56ae3fd9555082f0c)

File: test/elixir_search_extractor/search_keyword/search_keywords_test.exs [hunk 6]
```diff
@@ -1,10 +1,40 @@
...
+  import ElixirSearchExtractor.{AccountsFixtures, KeywordFixtures}
   alias ElixirSearchExtractor.SearchKeyword.Schemas.Keyword
   alias ElixirSearchExtractor.SearchKeyword.SearchKeywords
 
+  describe "paginated_user_keywords/2" do
+    test "returns user's keywords" do
+      user = user_fixture()
+      user_file = insert(:keyword_file, user: user)
+      user_keyword = insert(:keyword, keyword_file: user_file)
+      insert(:keyword)
```
Comment (hanam1ni) [inline test/elixir_search_extractor/search_keyword/search_keywords_test.exs]:
```suggestion
      _other_user_keyword = insert(:keyword)
```

**Review by hanam1ni — APPROVED** [review]

File: assets/css/screens/html-view.scss [hunk 7]
```diff
@@ -0,0 +1,7 @@
+.card-report {
+  height: 100%;
+}
+
+.html-view {
+  height: inherit;
+}
```
Comment (rosle) [inline assets/css/screens/html-view.scss]: For screen, we can create a body class 👍 

Add the helper to layout view
```elixir
# layout_view.ex
def body_class(conn) do
  "#{module_class_name(conn)} #{action_name(conn)}"
end

defp module_class_name(conn) do
  controller_module(conn)
  |> Phoenix.Naming.resource_name("Controller")
  |> String.replace("_", "-")
end
```

Use in the layout template
```html
<body class="<%= body_class(@conn) %>">
```

File: lib/elixir_search_extractor/search_keyword/queries/keyword_query.ex [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  import Ecto.Query
+  alias ElixirSearchExtractor.FileUpload.Schemas.KeywordFile
+  alias ElixirSearchExtractor.SearchKeyword.Schemas.Keyword
+
+  def user_keywords(user) do
+    from keyword in Keyword,
+      join: file in KeywordFile,
+      on: keyword.keyword_file_id == file.id,
+      where: file.user_id == ^user.id,
+      order_by: [desc: :inserted_at]
```
Comment (rosle) [inline lib/elixir_search_extractor/search_keyword/queries/keyword_query.ex]: For query, we prefer the pipe-based syntax if possible. The above query could be rewrite to:

```elixir
Keyword
|> join(:inner, [k], kf in assoc(k, :keyword_file)
|> where([_, kf], kf.user_id == ^user.id)
|> order_by([k, _], desc: k. inserted_at)
```

https://nimblehq.co/compass/development/code-conventions/elixir-phoenix/#database

File: test/elixir_search_extractor_web/features/keyword_page/view_html_view_page_test.exs [hunk 9]
```diff
@@ -0,0 +1,30 @@
+defmodule ElixirSearchExtractorWeb.KeywordPage.ViewHtmlViewPageTest do
+  use ElixirSearchExtractorWeb.FeatureCase, async: true
+
+  import ElixirSearchExtractor.AccountsFixtures
+
+  feature "view keyword html if keyword has html response", %{session: session} do
```
Comment (rosle) [inline test/elixir_search_extractor_web/features/keyword_page/view_html_view_page_test.exs]: `view` -> `views`

File: test/elixir_search_extractor_web/features/keyword_page/view_html_view_page_test.exs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    keyword = insert(:keyword, keyword_file: user_file, status: :completed, html: "Result View")
+
+    session
+    |> login(user_attributes.email, user_attributes.password)
+    |> visit(Routes.keyword_path(ElixirSearchExtractorWeb.Endpoint, :html_view, keyword.id))
+    |> focus_frame(Query.css(".html-view"))
+    |> assert_has(Query.text("Result View"))
+  end
+
+  feature "view No HTML view yet message if keyword status in precessing", %{session: session} do
```
Comment (rosle) [inline test/elixir_search_extractor_web/features/keyword_page/view_html_view_page_test.exs]:
```suggestion
  feature "views No HTML view yet message if keyword status in processing", %{session: session} do
```

**Comment by rafayet-monon** [comment]
@rosle Fixed your reviews in [26f1339](https://github.com/rafayet-monon/elixir_search_extractor/pull/41/commits/26f13396b18ca8c430bbfe4f6ca8af0617898bde)

File: assets/css/screens/_html-view.scss [hunk 11]
```diff
@@ -0,0 +1,9 @@
+body.html-view {
```
Comment (rosle) [inline assets/css/screens/_html-view.scss]: I think this file should be `_keyword.scss` following the controller name and then we can do

```scss
body.keyword.html-view {
```

Later if there is more actions, we can also separate files like

```
keyword
  - _html-view.scss
  - _index.scss
```

But for now I think just `_keyword.scss` works for us.

**Review by rosle — APPROVED** [review]
👍

**Pull request merged**